File: aurora-core/src/main/java/net/gree/aurora/domain/tablename/TableNameResolver.java
package net.gree.aurora.domain.tablename;

/**
 * ヒントから{@link TableName}を解決するためのインターフェイス。
 *
 * @param <T> ヒントの型
 */
public interface TableNameResolver<T> {

    /**
     * ヒントから{@link TableName}を解決する。
     *
     * @param hint ヒント
     * @return {@link TableName}
     */
    TableName apply(T hint);

}

File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/tablename/TableNameResolver.scala
package net.gree.aurora.scala.domain.tablename

import net.gree.aurora.domain.tablename.{TableNameResolver => JTableNameResolver}
import scala.language.implicitConversions

/**
 * [[net.gree.aurora.domain.tablename.TableNameResolver]]のScala版。
 *
 * @tparam T ヒントの型
 */
trait TableNameResolver[T] extends (T => TableName) {

  def apply(value: T): TableName

}

/**
 * コンパニオンオブジェクト。
 */
object TableNameResolver {

  /**
   * ScalaオブジェクトからJavaオブジェクトに変換する。
   *
   * @param self [[net.gree.aurora.scala.domain.tablename.TableNameResolver]]
   * @tparam T ヒントの型
   * @return [[net.gree.aurora.domain.tablename.TableNameResolver]]
   */
  implicit def toJava[T](self: TableNameResolver[T]) =
    new JTableNameResolver[T] {
      def apply(hint: T) = self(hint)
    }

}
